src/compiler/checker.ts

```diff
@@ -7571,6 +7574,78 @@ namespace ts {
+                        return current === (parent as CaseClause).expression;
+
+                    case SyntaxKind.BinaryExpression:
+                        const binaryExpr = parent as BinaryExpression;
+                        switch (binaryExpr.operatorToken.kind) {
+                            // Either operand of an equality/inequality comparison
+                            // should get a literal type.
+                            case SyntaxKind.EqualsEqualsEqualsToken:
+                            case SyntaxKind.ExclamationEqualsEqualsToken:
+                            case SyntaxKind.EqualsEqualsToken:
+                            case SyntaxKind.ExclamationEqualsToken:
+                                return current === binaryExpr.left || current === binaryExpr.right;
+
+                            case SyntaxKind.AmpersandAmpersandToken:
+                            case SyntaxKind.CommaToken:
+                                if (current === binaryExpr.right) {
+                                    continue loop;
+                                }
+                                break loop;
+
+                            case SyntaxKind.BarBarToken:
+                                if (current === binaryExpr.left || current === binaryExpr.right) {
+                                    continue loop;
+                                }
+                                break loop;
+                        }
+
+                        // No binary operators apply. Try to get the contextual type below.
+                        break loop;
+
+                    case SyntaxKind.ConditionalExpression:
+                        const conditional = parent as ConditionalExpression;
+                        if (current === conditional.whenTrue || current === conditional.whenFalse) {
+                            continue loop;
+                        }
+                        break loop;
+
+                    case SyntaxKind.ParenthesizedExpression:
+                        continue loop;
+
+                    default:
+                        // Nothing applies. Try to get the contextual type below.
+                        break loop;
+                }
+            }
+
+            // We haven't found a "literal match location" (i.e. a location that signals
+            // a literal should get a literal type). Check whether the contextual type
+            // of the literal has a literal type in it.
+            //
+            // You might ask why we're not passing 'current' or 'parent' into 'getContextualType'
+            // since we've already walked up several nodes anyway. This is not correct
+            // because several nodes *do not* strictly acquire a contextual type from their parents.
+            // An example of this is '||' expressions, whose right operand is contextually typed by
+            // its left operand if it could not acquire a contextual type from its parent.

```

Comment:
All of this would just go away if you make an `isEqualityComparisonOperand` method. Instead you'd keep it in `checkStringLiteralExpression`.